### ISSUE TITLE
Indexing torch.Tensor with pennylane.Wires object causes segfault

### ISSUE THREAD

**Issue by iamlucaswolf** [PennyLaneAI/pennylane#940]
#### Issue description

See title.

```python3
>>> from pennylane.wires import Wires
>>> import torch
>>> tensor = torch.arange(10)
>>> wires = Wires(0)
>>> tensor[wires]
Segmentation fault (core dumped)
```

The same code works fine with NumPy arrays, so my gut feeling is that the bug resides in PyTorch rather than PennyLane. I'll probably give this another look tomorrow and step through this with a debugger to make out the root cause.

I found this while working on #929, when modifying `pennylane._qubit_device.QubitDevice.sample()` to be agnostic w.r.t array/tensor-classes. AFAIK, this behaviour is not triggered anywhere in the current version; just be warned that this might happen.

* *Expected behavior:* (What you expect to happen)
N/A

* *Actual behavior:* (What actually happens)
See above

* *Reproduces how often:* (What percentage of the time does it reproduce?)
100% (on my machine)

* *System information:* (post the output of `import pennylane as qml; qml.about()`)
 Name: PennyLane
 Version: 0.12.0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: None
Author-email: None
License: Apache License 2.0
Location: /home/iamlucaswolf/.cache/pypoetry/virtualenvs/pennylane-9khiDOGD-py3.8/lib/python3.8/site-packages
Requires: toml, numpy, autograd, networkx, semantic-version, scipy, appdirs
Required-by: 
Platform info:           Linux-5.4.0-53-generic-x86_64-with-glibc2.27
Python version:          3.8.6
Numpy version:           1.19.4
Scipy version:           1.5.4
Installed devices:
    - default.gaussian (PennyLane-0.12.0)
    - default.mixed (PennyLane-0.12.0)
    - default.qubit (PennyLane-0.12.0)
    - default.qubit.autograd (PennyLane-0.12.0)
    - default.qubit.tf (PennyLane-0.12.0)
    - default.tensor (PennyLane-0.12.0)
    - default.tensor.tf (PennyLane-0.12.0)

**Comment by iamlucaswolf** [comment]
Having had a second look at this, it turned out that the reason for this behaviour stems from the `Wires` implementation:

In **PyTorch**, the indexed `torch.Tensor` object repeatedly calls `__getitem__()` on the indexing object, presumably until a valid index (integer, slices, ellipsis, None, or long/byte) Variable is returned. Since `Wires.__getitem__()` always returns another `Wires` object this never happens and the callstack grows until segfault.

In **NumPy**, the indexed `np.ndarray` eventually tries to call `__array__()` on the indexing object, and since `Wires` has a `__array__` implementation everything works fine.


For #929, I replaced replaced occurences of `array[wires]` by `array[wires.tolist()]` which works for both NumPy and PyTorch arrays/tensors. Still, I believe it would be cleaner if this would be fixed in `Wires` directly.

**Comment by josh146** [comment]
Thanks for catching this @iamlucaswolf!

> For #929, I replaced replaced occurences of array[wires] by array[wires.tolist()] which works for both NumPy and PyTorch arrays/tensors.

I think this is a reasonable short term fix.

> Still, I believe it would be cleaner if this would be fixed in Wires directly.

I'm still trying to gather my thoughts, but I'm not sure I'm 100% clear what a long term fix in the Wires class looks like, since this seems to be a peculiarity with how indexing is performed in PyTorch; I wonder if there is a way to treat array-like objects as valid indices (similarly, I wonder if TF respects the `__array__` method when indexing).

A more breaking solution would be to have the Wires class simply return lists when indexed, but I remember @mariaschuld making strong arguments that the Wires class, when indexed, should always return another Wires object.

**Comment by mariaschuld** [comment]
Thanks @iamlucaswolf for digging so deep! :)  

While I see that it is convenient to just index with a wires object that happens to represent integer-valued wire labels, I think this should not be encouraged. Wires could also be a sequence of the form `['a', 0, 'qubit3`]`, in which indexing does not make sense. Since the class is not user-exposed, and PennyLane should always be agnostic as to what wire labels a user/device chooses, I cannot think of a case in which something like `tensor[wires]` could be safely used. 

Take `QubitDevice` for example, as you mention: While by default, the device base class uses a map that [translates the user's wire labels to integer-valued labels](https://github.com/PennyLaneAI/pennylane/blob/fe32d314a76570b160494d8c31fca677dbdc44f7/pennylane/_device.py#L186) (like `Wires=[0, 1, 3])`, a device can easily overwrite this, in which case other methods based on indexing fail.

In other words, the inconvenience of not being able to index is a bit of a safety feature!

Would it be useful to raise an error when using wires for indexing? Is there an elegant way to actually do this in the wires class (since it does not know what calls its `__getitem__`? Or shall I just add a warning to the documentation?

**Comment by antalszava** [comment]
Closing this as using `Wires` objects for indexing is not supported. Feel free to leave a comment for further thoughts/points.
